proposals/0512-nofieldselectors-per-datatype.md

```diff
@@ -0,0 +1,203 @@
+
+## Costs and Drawbacks
+
+GHC already has a facility for deciding on whether or not to generate field selectors for a datatype.
+Extending this facility to check for a datatype local override should not be particularly thorny.
+
+An additional annotation is more syntax to learn.
+The `{-# FieldSelectors #-}` syntax is a bit ugly, though it is consistent with other annotations.
+
+Implementing this proposal may spark desire for further datatype annotations.
+The `StrictData` extension comes to mind immediately - being able to write `data X = X {-# Strict #-}` seems like an obvious application.
+
+If the alternative for *field-level* annotations is appealing, then we will need to consider what syntax we may want for this.
+`StrictData` annotations on a per-field basis are done with `BangPatterns` (and a laziness annotation can be provided with `~` as a prefix operator).
+Using a `{-# Blah #-}` comment as a field annotation introduces a second means of annotating a field, and in particular, it is somewhat noisy.
+If we're able to annotate fields, we may also want the ability to say that some fields are "private" or "public" much like other object and record systems permit.
+Finally, field annotations could potentially be used in `DerivingVia` - see [this Reddit post by `Iceland_jack`](https://www.reddit.com/r/haskell/comments/pbq9rn/via_fields_finer_granularity_in_deriving/).
+
+## Alternatives
+
+There is currently no way to get record selectors for record construction and record pattern matching without record syntax.
+
+For `OverloadedRecordDot` field access, synthetic `HasField` instances can be defined.
+
+Another potential extension of this design is to apply it to *fields* as well as a type.
+The `StrictData` analogy is already implemented here with `BangPatterns`.
+
+```haskell
+-- This datatpye generates the `age` field selector, but not the `name`
+-- selector.
+data Foo = Foo
+    { {-# NoFieldSelector #-} name :: String 
+    , age :: Int
+    }
+```
+
+```haskell
+{-# language NoFieldSelectors #-}
+
+module X where
+
+data Point = Point
+    { x :: Int
+    , y :: Int
+    , {-# FieldSelector #-} z :: Int 
+    }
+```
+
+Syntactically, the `{-# Blah #-}` style is lifted from `{-# Overlapping #-}`.
+However, it's pretty ugly and verbose.
+A further extension could introduce new keywords that are only used in record declaration syntax with the same meaning, allowing you to write, eg,
+
+```haskell
+data Point = Point
+    { x :: strict selector Int
+    , y :: lazy field Int
+    }
+```
+
+## Unresolved Questions
+
+Since this is a specialization of an existing language behavior, I don't believe there is anything to be resolved here.
+
+
+## Implementation Plan
+
+The datatype for a record declaration is extended with a field `Maybe FieldSelectorOverride`.
+When GHC is deciding whether or not to generate field selectors, this field is checked as a final source of truth.
+
+This change is also propagated to `template-haskell`.
```

Comment:
Can you already say whether you plan to implement this yourself? Or know who will?

Reply:
I've assigned the implementation to @wavewave, Mercury's GHC engineer, if it gets approved/accepted.